src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -181,12 +189,20 @@ bool MulticopterLandDetector::_get_ground_contact_state()
 
 	// if we have a valid velocity setpoint and the vehicle is demanded to go down but no vertical movement present,
 	// we then can assume that the vehicle hit ground
-	_in_descend = _is_climb_rate_enabled()
-		      && (_vehicle_local_position_setpoint.vz >= land_speed_threshold);
+	_in_descend = _is_climb_rate_enabled() && (_vehicle_local_position_setpoint.vz >= land_speed_threshold);
 	bool hit_ground = _in_descend && !vertical_movement;
 
+	bool ground_contact = false;
+
+	if (_maybe_landed_hysteresis.get_state() || _landed_hysteresis.get_state()) {
```

Comment:
I was actually hoping @RomanBapst could explain that one. If it wasn't obvious I resurrected this from his old PR.

Reply:
@BazookaJoe1900 I don't see a reason why it should not work.
Anyway I tested this again today and it does not work but now for another reason. It seems like during landing the flighttask is sometimes publishing NAN for vz_setpoint which make the _hit_ground condition false and so landing is never detected.

The one and only reason why I made this change in the past was to reduce the chance of false positives in ground contact detection.

Reply:
This is in response to a false positive that caused a crash. https://github.com/PX4/Firmware/issues/13124#issuecomment-642002013

Reply:
I've added a vz setpoint finite check.

Reply:
it took me some time to understand but now I get it.
In order to enter to ground contact in this logic you have similar conditions as before. (_has_low_thrust() && hit_ground), but with AND.
but after it valid for enough time that cause _maybe_landed_hysteresis to be valid, you only need _has_low_thrust() || hit_ground. and that the same conditions as before.

that a big change of the logic. because entering to ground_contact can be much 'harder', and delayed. has this been tested on several vehicles?
Did I get it right?

and still, if _maybe_landed_hysteresis is true, _landed_hysteresis is true for sure so the || is not needed.

Reply:
I'm not sure if this makes sense because it means that the `_get_ground_contact_state` now depends on `_maybe_landed` and `_landed`, however, these are handled "outside" as I understand it:
https://github.com/PX4/Firmware/blob/326d8efc166331d5b9a0fc96836b68e82142a409/src/modules/land_detector/LandDetector.cpp#L76-L81

Ok, so the way it works is that `ground_contact` leads to the controller reducing thrust to idle here:
https://github.com/PX4/Firmware/blob/d92e66863ac057a4a691d482b8566081700e9e1e/src/modules/mc_pos_control/mc_pos_control_main.cpp#L644-L647
and that then triggers landed eventually (or freefall worst case).

So maybe we could add a check in the position controller to check both `ground_contact` AND `maybe_landed` and only then start reducing thrust.

Reply:
> So maybe we could add a check in the position controller to check both ground_contact AND maybe_landed and only then start reducing thrust.

Few notes:
- the states ground_contact, maybe_landed, and landed are not really independent states, each one depend on the previous, so no need to check both ground_contact AND maybe_landed, just maybe_landed

- This is something we discussed before. to reduce trust on later stage. but that might lead to flips because of late detection of ground. 

- my proposal is that "between"  ground_contact to maybe_landed the thrust will not cut to zero, but gradually reduced. so it both "safe" in the matter of quick response to detection of ground, but if it was false detection, it probably be safer to recover from.

Reply:
The one and only reason why I made this change in the past was to reduce the chance of false positives in ground contact detection.

> ...because entering to ground_contact can be much 'harder', and delayed

That's the goal (to eliminate the false positives). Then the idea is to then make use of the hover thrust estimate to tighten the low thrust threshold and minimize any potential negative impact this change might introduce for existing "good" setups.


> my proposal is that "between" ground_contact to maybe_landed the thrust will not cut to zero, but gradually reduced. so it both "safe" in the matter of quick response to detection of ground, but if it was false detection, it probably be safer to recover from.

Let's be sure to come back to this. @MaEtUgR do you have an issue or anything for this idea?